File: 0156. Binary Tree Upside Down/main0156.py

class Solution0156(object):
    def upsideDownBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if (not root) or (not root.left):
            return root
        newRoot = self.upsideDownBinaryTree(root.left)
        root.left.right = root
        root.left.left = root.right
        root.left = None
        root.right = None
        return newRoot

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

sol = Solution0156()
root =  TreeNode(1)
root.left =  TreeNode(2)
root.right =  TreeNode(3)
cur = root
cur = cur.left
cur.left =  TreeNode(4)
cur.right =  TreeNode(5)        
res = sol.upsideDownBinaryTree(root)
print(root.val)
print(res.val)

File: 0156. Binary Tree Upside Down/main0156.java
public class main0156 {
    public static void main(String[] args) {
        Solution0156 sol = new Solution0156();
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        TreeNode cur = root;
        cur = cur.left;
        cur.left = new TreeNode(4);
        cur.right = new TreeNode(5);        
        TreeNode res = sol.upsideDownBinaryTree(root);
        System.out.println(root.val);
        System.out.println(res.val);
    }
}

class Solution0156 {
    public TreeNode upsideDownBinaryTree(TreeNode root) {
        if (root == null || root.left == null) {
            return root;
        }
        TreeNode newRoot = upsideDownBinaryTree(root.left);
        root.left.right = root;
        root.left.left = root.right;
        root.left = null;
        root.right = null;
        
        return newRoot;
    }
}


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
